source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/locale-input/field.tsx

```diff
@@ -0,0 +1,141 @@
+import * as React from 'react';
+import { useMemo, memo } from 'react';
+import Autosuggest from 'react-autosuggest';
+import { assert } from '@orbit/utils';
+
+import { withTranslations, i18nProps, useLdml } from '~/lib/i18n';
+
+import { filterForValidAttributes } from '~/lib/dom';
+
+import { getSuggestions, findLanguageCode, sortComparer, findLanguageInfo } from './-utils/helpers';
+import { localizeTagData } from './-utils/localize';
+import { Suggestion } from './suggestion';
+
+export interface IProps {
+  value: string; // tag
+  onChange: (localeCode: string) => void;
+  data: ILanguageInfo[];
+}
+
+interface IState {
+  value?: string;
+  isMatch: boolean;
+  suggestions: ILanguageInfo[];
+  languageInfo: ILanguageInfo;
+}
+
+class Field extends React.Component<IProps & i18nProps, IState> {
+  getSuggestions: (value: string) => ILanguageInfo[];
+  findLanguageCode: (value: string) => string;
+  findLanguageInfo: (value: string) => ILanguageInfo;
+
+  constructor(props) {
+    super(props);
+
+    assert(`The Locale Picker Field needs a data set. The passed data set is empty.`, props.data);
+
+    this.getSuggestions = getSuggestions(props.data).bind(this);
+    this.findLanguageCode = findLanguageCode(props.data);
+    this.findLanguageInfo = findLanguageInfo(props.data);
+
+    const value = this.findLanguageCode(props.value) || '';
```

Comment:
what would happen if `findLanguageCode` is not found and `value` is empty? should we throw here or will this behave properly?

Reply:
also should `findLanguageCode` encapsulate the logic around returning empty when not found?

Reply:
findLanguageCode just does `Array#find`, so it'd return undefined if there is no result. if value is empty, then undefined will also be returned, as every value it searches over isn't `undefined`. (this will behave properly)

I don't think findLanguageCode should encapsulate the `|| ''` because the `|| ''` is specifically to force the input in the template of this component to be a "controlled" component. In the event that an input's value is undefined, the input is "uncontrolled", and React yells at you for converting it from uncontrolled to controlled at runtime. It's silly, cause `undefined` _can't_ be a value of an input field, and I think the `|| ''` should actually be a JSX concern :-\